### PR TITLE
PLAT-124934: [Agate Refactoring] Sync Picker, RangePicker, DatePicker, TimePicker, DateTime, and DateComponentPicker with Sandstone

### PR DESCRIPTION
**Pull request by daniel-stoian-lgp**
Enact-DCO-1.0-Signed-off-by: Daniel Stoian daniel.stoian@lgepartner.com

### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
DatePicker:
- added `onSpotlightDissapear` and `spotlightDisabled` props
- applied `wrap` to internal DateTimeComponentPickers
- added unit tests
- updated screenshot tests

TimePicker:
- added `onSpotlightDissapear` and `spotlightDisabled` props
- applied `wrap` to internal DateTimeComponentPickers, except for Meridiem
- added unit tests
- updated screenshot tests

DateTimePicker
- updated screenshot tests

Picker:
- added `noAnimation`, and `wrap` props
- fixed it to not use internal/Picker/PickerDecorator because reverseTransition information was lost
- added unit test
- updated screenshot tests

RangePicker: 
- added `noAnimation`, and `wrap` props
- wraped children in `agate/internal/Picker/PickerItem`
- updated screenshot tests

internal/DateComponentPicker
- changed the usage of `Picker` to `internal/Picker`
- wrapped children into `PickerItem`

internal/DateComponentRangePicker
-  added `accesabilityHint`, and `wrap` props
- wrapped children into `PickerItem`

internal/DateTime
- removed props `locale` and `value`

internal/Picker:
- added `aria-label`, `onSpotlightDissapear`, `spotlightDisabled` and `wrap`,  props
- updated defaultProps

### Links
[//]: # (Related issues, references)
PLAT-124934

### Comments

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,12 +3,21 @@
 The following is a curated list of changes in the Enact agate module, newest changes on the top.
 
 ## [unreleased]
+
+### Added
+- `agate/DatePicker` prop `onSpotlightDissapear` and `spotlightDisabled`
+- `agate/internal/DateComponentRangePicker` prop `accesabilityHint` and `wrap`
+- `agate/internal/Picker` prop `onSpotlightDissapear`, `spotlightDisabled` and `wrap`
```
Comment (stanca-pop-lgp) [inline CHANGELOG.md:10]: Cam you move these at the bottom of the list?

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +28,10 @@ The following is a curated list of changes in the Enact agate module, newest cha
 - `agate/Picker` picker item width in horizontal for silicon
 - `agate/TemperatureControl` to not be draggable when it's disabled
 
+### Removed
+- `agate/internal/DateTime` prop `locale` and `value`
+
+
```
Comment (stanca-pop-lgp) [inline CHANGELOG.md:34]: I don't think the 2 empty lines are necessary.

**Review by stanca-pop-lgp — APPROVED** [review]
LGTM. Great job!

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,14 +3,25 @@
...
 ## [unreleased]
+
+### Added
+
+- `agate/DatePicker` prop `onSpotlightDissapear` and `spotlightDisabled`
+- `agate/Picker` prop `noAnimation` and `wrap`
+- `agate/RangePicker` prop `noAnimation` and `wrap`
+- `agate/TimePicker` prop `onSpotlightDissapear` and `spotlightDisabled`
+- `agate/internal/DateComponentRangePicker` prop `accesabilityHint` and `wrap`
+- `agate/internal/Picker` prop `onSpotlightDissapear`, `spotlightDisabled` and `wrap`
```
Comment (MikyungKim) [inline CHANGELOG.md:14]: We don't write change logs for internal components.

File: CHANGELOG.md [hunk 4]
```diff
@@ -3,14 +3,25 @@
 The following is a curated list of changes in the Enact agate module, newest changes on the top.
 
 ## [unreleased]
+
+### Added
+
+- `agate/DatePicker` prop `onSpotlightDissapear` and `spotlightDisabled`
```
Comment (MikyungKim) [inline CHANGELOG.md:9]:
```suggestion
- `agate/DatePicker` prop `onSpotlightDisappear` and `spotlightDisabled`
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -3,14 +3,25 @@
...
+- `agate/Picker` prop `noAnimation` and `wrap`
+- `agate/RangePicker` prop `noAnimation` and `wrap`
+- `agate/TimePicker` prop `onSpotlightDissapear` and `spotlightDisabled`
```
Comment (MikyungKim) [inline CHANGELOG.md:12]:
```suggestion
- `agate/TimePicker` prop `onSpotlightDisappear` and `spotlightDisabled`
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,14 +3,25 @@
...
  
 ### Fixed
+
 - `agate/DateTimePicker` returned value by onChange event
 - `agate/FanSpeedControl` to center text when there is no icon
 - `agate/Heading` to support `spacing` for Gallium and Silicon
 - `agate/IncrementSlider` button color for Gallium skin
 - `agate/Keypad` to not show console error in sampler
 - `agate/Panels` to show close button properly for night mode
+- `agate/Picker` to not use internal/Picker/PickerDecorator
```
Comment (MikyungKim) [inline CHANGELOG.md:24]: I don't think we need to mention this.

File: CHANGELOG.md [hunk 7]
```diff
@@ -22,6 +33,10 @@ The following is a curated list of changes in the Enact agate module, newest cha
 - `agate/Picker` picker item width in horizontal for silicon
 - `agate/TemperatureControl` to not be draggable when it's disabled
 
+### Removed
+
+- `agate/internal/DateTime` prop `locale` and `value`
```
Comment (MikyungKim) [inline CHANGELOG.md:38]: Let's remove internal things.

File: DatePicker/DatePickerBase.js [hunk 8]
```diff
@@ -164,6 +173,15 @@ const DatePickerBase = kind({
 		 */
 		rtl: PropTypes.bool,
 
+		/**
+		 * Disables 5-way spotlight from navigating into the component.
+		 *
+		 * @type {Boolean}
+		 * @default false
```
Comment (MikyungKim) [inline DatePicker/DatePickerBase.js:180]: We don't need to add this for default value false.

File: DatePicker/DatePickerBase.js [hunk 9]
```diff
@@ -179,7 +197,8 @@ const DatePickerBase = kind({
 	defaultProps: {
 		maxYear: 2099,
 		minYear: 1900,
-		disabled: false
+		disabled: false,
+		spotlightDisabled: false
```
Comment (MikyungKim) [inline DatePicker/DatePickerBase.js:201]: Let's remove this since we don't need it.

File: Picker/Picker.js [hunk 10]
```diff
@@ -64,21 +75,29 @@ const PickerBase = kind({
...
+		 * @type {Boolean}
+		 * @default false
+		 * @public
+		 */
+		wrap: PropTypes.bool
 	},
 
 	defaultProps: {
-		value: 0
+		orientation: 'vertical',
```
Comment (MikyungKim) [inline Picker/Picker.js:92]: Let's move the definition of propTypes for `orientation` from below to above.

File: Picker/Picker.js [hunk 11]
```diff
@@ -64,21 +75,29 @@ const PickerBase = kind({
...
 		 */
-		value: PropTypes.number
+		value: PropTypes.number,
+
+		/**
+		 * Allows picker to continue from the start of the list after it reaches the end and
+		 * vice-versa.
+		 *
+		 * @type {Boolean}
+		 * @default false
```
Comment (MikyungKim) [inline Picker/Picker.js:85]: Don't need to define.

File: Picker/Picker.js [hunk 12]
```diff
@@ -64,21 +75,29 @@ const PickerBase = kind({
...
+		 * @public
+		 */
+		wrap: PropTypes.bool
 	},
 
 	defaultProps: {
-		value: 0
+		orientation: 'vertical',
+		value: 0,
+		wrap: false
```
Comment (MikyungKim) [inline Picker/Picker.js:94]: Please remove this.

File: Picker/Picker.js (hunk 10 above)
Comment (MikyungKim) [reply Picker/Picker.js:92]: Actually, you could move the all propTypes definition below to propTypes block.
We don't use it directly but we do pass the prop with `rest`.

File: RangePicker/RangePicker.js [hunk 13]
```diff
@@ -120,7 +141,7 @@ const RangePickerBase = kind({
 	render: ({value, ...rest}) => {
 		return (
 			<PickerCore {...rest} index={0} value={value}>
```
Comment (MikyungKim) [inline RangePicker/RangePicker.js:143]: I think we are missing `type="number"` here.

File: RangePicker/RangePicker.js [hunk 14]
```diff
@@ -107,7 +119,16 @@ const RangePickerBase = kind({
...
+		step: PropTypes.number,
+
+		/**
+		 * Allows picker to continue from the start of the list after it reaches the end and
+		 * vice-versa.
+		 *
+		 * @type {Boolean}
+		 * @public
+		 */
+		wrap: PropTypes.bool
```
Comment (MikyungKim) [inline RangePicker/RangePicker.js:131]: Would you please move the propTypes definition down below to here?

File: TimePicker/TimePickerBase.js [hunk 15]
```diff
@@ -195,17 +195,36 @@ const TimePickerBase = kind({
...
 		 * @private
 		 */
-		rtl: PropTypes.bool
+		rtl: PropTypes.bool,
+
+		/**
+		 * Disables 5-way spotlight from navigating into the component.
+		 *
+		 * @type {Boolean}
+		 * @default false
```
Comment (MikyungKim) [inline TimePicker/TimePickerBase.js:219]: Don't need it.

File: TimePicker/TimePickerBase.js [hunk 16]
```diff
@@ -195,17 +195,36 @@ const TimePickerBase = kind({
...
+		 * @public
+		 */
+		spotlightDisabled: PropTypes.bool
 	},
 
 	defaultProps: {
-		disabled: false
+		disabled: false,
+		spotlightDisabled: false
```
Comment (MikyungKim) [inline TimePicker/TimePickerBase.js:227]: We don't need it.

File: internal/DateTime/DateTime.js [hunk 17]
```diff
@@ -18,9 +18,7 @@ const DateTimeBase = kind({
 	name: 'DateTime',
 
 	propTypes:  /** @lends agate/internal/DateTime.DateTime.prototype */ {
-		css: PropTypes.object,
-		locale: PropTypes.string,
-		value: PropTypes.any
+		css: PropTypes.object
```
Comment (MikyungKim) [inline internal/DateTime/DateTime.js:21]: Would you please revert the change in this file? Since we added the prop and delete it from https://github.com/enactjs/agate/pull/357 intentionally. Without this, the DOM will have `locale` and `value` property that we don't want.

File: internal/Picker/Picker.js [hunk 18]
```diff
@@ -217,6 +247,15 @@ const PickerBase = kind({
 		 */
 		skin: PropTypes.string,
 
+		/**
+		 * When `true`, the component cannot be navigated using spotlight.
+		 *
+		 * @type {Boolean}
+		 * @default false
```
Comment (MikyungKim) [inline internal/Picker/Picker.js:254]: Please remove default value.

File: internal/Picker/Picker.js [hunk 19]
```diff
@@ -253,14 +292,27 @@ const PickerBase = kind({
...
+		 * @public
+		 */
+		wrap: PropTypes.bool
 	},
 
 	defaultProps: {
 		accessibilityHint: '',
 		orientation: 'vertical',
+		reverseTransition: false,
```
Comment (MikyungKim) [inline internal/Picker/Picker.js:311]: Please remove default false props

File: RangePicker/RangePicker.js (hunk 13 above)
Comment (MikyungKim) [reply RangePicker/RangePicker.js:143]: Hm.. Think we don't have `type` prop for `internal/Picker`.

File: internal/Picker/Picker.js [hunk 20]
```diff
@@ -389,6 +451,46 @@ const PickerBase = kind({
 			}
 		};
 
+		const currentItemIndex = () => {
```
Comment (MikyungKim) [inline internal/Picker/Picker.js:454]: Can we make the functions in render to computed, please?

File: DatePicker/DatePickerBase.js [hunk 21]
```diff
@@ -164,6 +173,15 @@ const DatePickerBase = kind({
...
 		rtl: PropTypes.bool,
 
+		/**
+		 * Disables 5-way spotlight from navigating into the component.
+		 *
+		 * @type {Boolean}
+		 * @default false
+		 * @public
+		 */
+		spotlightDisabled: PropTypes.bool,
```
Comment (MikyungKim) [inline DatePicker/DatePickerBase.js:183]: Could you explain why did you add this prop?

File: TimePicker/TimePickerBase.js [hunk 22]
```diff
@@ -195,17 +195,36 @@ const TimePickerBase = kind({
...
+		rtl: PropTypes.bool,
+
+		/**
+		 * Disables 5-way spotlight from navigating into the component.
+		 *
+		 * @type {Boolean}
+		 * @default false
+		 * @public
+		 */
+		spotlightDisabled: PropTypes.bool
```
Comment (MikyungKim) [inline TimePicker/TimePickerBase.js:222]: Also, why do we need this?

File: tests/screenshot/apps/components/Picker.js [hunk 23]
```diff
@@ -1,10 +1,41 @@
...
-	<Picker disabled>{[]}</Picker>,
-	<Picker orientation="horizontal">{[]}</Picker>
+	<Picker>{pickerList.temperatures}</Picker>,
+	<Picker disabled>{pickerList.temperatures}</Picker>,
+	<Picker value={1}>{pickerList.temperatures}</Picker>,
+	<Picker orientation="horizontal">{pickerList.temperatures}</Picker>,
+	// RTL
+	{
+		locale: 'ar-SA',
+		component: 	<Picker>{pickerList.temperatures}</Picker>
```
Comment (MikyungKim) [inline tests/screenshot/apps/components/Picker.js:23]:
```suggestion
		component: <Picker>{pickerList.temperatures}</Picker>
```

File: tests/screenshot/apps/components/Picker.js [hunk 24]
```diff
@@ -1,10 +1,41 @@
...
+	},
+	{
+		locale: 'ar-SA',
+		component: 	<Picker disabled>{pickerList.temperatures}</Picker>
```
Comment (MikyungKim) [inline tests/screenshot/apps/components/Picker.js:27]:
```suggestion
		component: <Picker disabled>{pickerList.temperatures}</Picker>
```

File: tests/screenshot/apps/components/Picker.js [hunk 25]
```diff
@@ -1,10 +1,41 @@
...
+	},
+	{
+		locale: 'ar-SA',
+		component: 	<Picker orientation="horizontal">{pickerList.temperatures}</Picker>
```
Comment (MikyungKim) [inline tests/screenshot/apps/components/Picker.js:31]:
```suggestion
		component: <Picker orientation="horizontal">{pickerList.temperatures}</Picker>
```

**Review by MikyungKim — CHANGES_REQUESTED** [review]
Please find my comments. And also, I found weird behavior on `wrap` in TimePicker.
I'll attach the movie to the ticket. Thank you!

File: RangePicker/RangePicker.js (hunk 13 above)
Comment (daniel-stoian-lgp) [reply RangePicker/RangePicker.js:143]: I will add it and make calculations based on it

File: TimePicker/TimePickerBase.js (hunk 22 above)
Comment (daniel-stoian-lgp) [reply TimePicker/TimePickerBase.js:222]: well, to disable 5-way navigation on the component. it is not clearly visible on Gallium, but it is on Silicon

File: DatePicker/DatePickerBase.js (hunk 21 above)
Comment (daniel-stoian-lgp) [reply DatePicker/DatePickerBase.js:183]: to disable 5-way navigation on the component. Focus state is not clearly visible on Gallium, but it is on Silicon.

File: CHANGELOG.md [hunk 26]
```diff
@@ -7,7 +7,11 @@ The following is a curated list of changes in the Enact agate module, newest cha
 ### Added
 
 - `agate/Button` prop `iconOnly`, `iconPosition` and `minWidth`
+- `agate/DatePicker` prop `onSpotlightDisappear` and `spotlightDisabled`
```
Comment (MikyungKim) [inline CHANGELOG.md:10]: I'm wondering should we add these props to `DateTimePicker` too?

File: DatePicker/DatePickerBase.js (hunk 21 above)
Comment (MikyungKim) [reply DatePicker/DatePickerBase.js:182]: Oh... I see..

**Review by MikyungKim — COMMENTED** [review]
Good work! I left some concerns here and the ticket. Please find it. Thank you!

File: CHANGELOG.md (hunk 26 above)
Comment (daniel-stoian-lgp) [reply CHANGELOG.md:10]: good idea

File: DateTimePicker/DateTimePicker.js [hunk 27]
```diff
@@ -64,15 +81,15 @@ const DateTimePickerBase = kind({
...
 	},
 
-	render ({disabled, onChange, value, ...rest}) {
+	render ({disabled, onChange, spotlightDisabled, value, ...rest}) {
 
 		return (
 			<Row {...rest} className={css.dateTimePicker} align="center center">
 				<Row align="center center">
-					<TimePicker css={css} disabled={disabled} onChange={onChange} value={value} />
+					<TimePicker css={css} disabled={disabled} onChange={onChange} spotlightDisabled={spotlightDisabled} value={value} />
```
Comment (MikyungKim) [inline DateTimePicker/DateTimePicker.js:89]: Shouldn't `onSpotlightDisappear` be here and there, also?

Comment (daniel-stoian-lgp) [reply DateTimePicker/DateTimePicker.js]: upps. fixed

**Review by MikyungKim — APPROVED** [review]
LGTM!
